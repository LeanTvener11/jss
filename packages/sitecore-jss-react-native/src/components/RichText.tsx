import React from 'react';
import PropTypes from 'prop-types';
import { Dimensions } from 'react-native';
import RenderHtml from 'react-native-render-html';

export interface RichTextProps {
  [htmlViewProps: string]: unknown;
  /** The rich text field data. */
  field: {
    value?: string;
    editable?: string;
  } | null;
}

export const RichText: React.SFC<RichTextProps> = ({ field, ...otherProps }) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  const htmlValue = field.editable ? field.editable : field.value;

  return (
    <RenderHtml
      contentWidth={Dimensions.get('window').width}
      source={{ html: htmlValue }}
      {...otherProps}
    />
  );
};

RichText.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.string,
    editable: PropTypes.string,
  }),
};

RichText.displayName = 'RichText';
