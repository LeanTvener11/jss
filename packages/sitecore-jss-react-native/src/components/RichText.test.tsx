import React from 'react';
import renderer from 'react-test-renderer';
import { RichText } from './RichText';

describe.only('<RichText />', () => {
  test('should render nothing with missing field', () => {
    const field: unknown = null;
    const rendered = renderer.create(<RichText field={field} />);
    expect(rendered).toMatchSnapshot();
  });

  test('should render nothing with missing editable and value', () => {
    const props = {
      field: {},
    };
    const rendered = renderer.create(<RichText {...props} />);
    expect(rendered).toMatchSnapshot();
  });

  test('should render editable value if present', () => {
    const props = {
      field: {
        value: 'this value should not be rendered',
        editable: '<p>this value should be rendered &copy;</p>',
      },
    };
    const rendered = renderer.create(<RichText {...props} />);

    expect(rendered).toMatchSnapshot();
  });

  test('should render value if editable is undefined', () => {
    const props = {
      field: {
        value: 'this value should be rendered',
        editable: '',
      },
    };
    const rendered = renderer.create(<RichText {...props} />);

    expect(rendered).toMatchSnapshot();
  });

  it('should render other attributes with other props provided', () => {
    const field = {
      editable: 'value with <p>markup</p>',
    };

    // 'baseStyle' prop will be applied to the html View (outer container).
    // https://meliorence.github.io/react-native-render-html/api/renderhtmlprops
    const rendered = renderer.create(
      <RichText
        field={field}
        baseStyle={{ flex: 1 }}
      />
    );
    expect(rendered).toMatchSnapshot();
  });
});
