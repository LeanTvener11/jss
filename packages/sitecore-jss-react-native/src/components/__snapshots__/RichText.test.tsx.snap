// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichText /> should render editable value if present 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
      undefined,
    ]
  }
  testID="html"
>
  <View
    style={
      Array [
        Object {},
        undefined,
        undefined,
      ]
    }
    testID="body"
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 14,
            "marginTop": 14,
          },
          undefined,
          undefined,
        ]
      }
      testID="p"
    >
      <Text
        allowFontScaling={true}
        selectable={false}
        style={
          Array [
            Object {
              "fontSize": 14,
            },
            undefined,
            undefined,
          ]
        }
      >
        this value should be rendered Â©
      </Text>
    </View>
  </View>
</View>
`;

exports[`<RichText /> should render nothing with missing editable and value 1`] = `null`;

exports[`<RichText /> should render nothing with missing field 1`] = `null`;

exports[`<RichText /> should render other attributes with other props provided 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
      undefined,
    ]
  }
  testID="html"
>
  <View
    style={
      Array [
        Object {},
        undefined,
        undefined,
      ]
    }
    testID="body"
  >
    <Text
      allowFontScaling={true}
      selectable={false}
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          undefined,
          undefined,
        ]
      }
    >
      value with
    </Text>
    <View
      style={
        Array [
          Object {
            "marginBottom": 14,
            "marginTop": 14,
          },
          undefined,
          undefined,
        ]
      }
      testID="p"
    >
      <Text
        allowFontScaling={true}
        selectable={false}
        style={
          Array [
            Object {
              "fontSize": 14,
            },
            undefined,
            undefined,
          ]
        }
      >
        markup
      </Text>
    </View>
  </View>
</View>
`;

exports[`<RichText /> should render value if editable is undefined 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
      undefined,
    ]
  }
  testID="html"
>
  <View
    style={
      Array [
        Object {},
        undefined,
        undefined,
      ]
    }
    testID="body"
  >
    <Text
      allowFontScaling={true}
      selectable={false}
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          undefined,
          undefined,
        ]
      }
    >
      this value should be rendered
    </Text>
  </View>
</View>
`;
